test/scripts/products/h24-verify-selfmanaged-apicast-and-custom-policy/h24-verify-selfmanaged-apicast-and-custom-policy.go

```diff
@@ -0,0 +1,386 @@
+	command := "PASSWORD=" + adminPassword + " DEDICATED_ADMIN_PASSWORD=" + adminPassword + " " + basepath + "scripts/setup-sso-idp.sh"
+	err := run_shell_command(command, "create_customer_users()")
+	return err
+}
+
+// Create a new namespace (selfmanaged-apicast) for self-managed APIcast (SOP item 2)
+func create_new_namespace() error {
+	log.Info("create_new namespace()")
+	exists, err := check_if_namespace_exists(apicast_namespace)
+	if err != nil {
+		return err
+	}
+	if !exists {
+		log.Info("Namespace creating")
+		command := "oc new-project " + apicast_namespace
+		err = run_shell_command(command, "namespace()")
+		return err
+	}
+	return nil
+}
+
+func check_if_namespace_exists(namespace string) (bool, error) {
+	log.Info("check_if_namespace_exists()")
+	command := "oc projects |grep " + namespace
+	log.Info("Command: " + command)
+	out, _ := exec.Command("sh", "-c", command).Output()
+	log.Info("Out from grep namespace: " + string(out))
+	if strings.Contains(string(out), namespace) {
+		log.Info("Namespace exists")
+		return true, nil
+	}
+	log.Info("Namespace dose not exists")
+	return false, nil
+}
+
+// Import APIcast Image (SOP item 3)
+func import_apicast_image(tag string) error {
+	log.Info("import_apicast_image")
+	//command: oc import-image 3scale-amp2/apicast-gateway-rhel8:<tag> --from=registry.redhat.io/3scale-amp2/apicast-gateway-rhel8:<tag> --confirm
+	command := "oc import-image 3scale-amp2/apicast-gateway-rhel8:" + tag + " --from=registry.redhat.io/3scale-amp2/apicast-gateway-rhel8:" + tag + " --confirm"
+	err := run_shell_command(command, "import_apicast_image()")
+	return err
+}
+
+// step 4. Create a secret with credentials to `registry.redhat.io`.
+// -- NOT used in current script version --
+// SKIP if step 3 is ok
+// This function is not in use meanwhile (not verified), as
+// OSD clusters are allowed to access that registry by default
+func create_redhatio_secret(user, passwd, email string) error {
+	log.Info("create_redhatio_secret()")
+	command := "oc creat secret docker-registry redhatio --docker-server=registry.redhat.io --docker-username=" + user + " --docker-password=" + passwd + " --docker-email=" + email
+	err := run_shell_command(command, "create_redhatio_secret()")
+	return err
+}
+
+// step 5. Configure the pull secret.
+// -- NOT used in current script version --
+// SKIP if step 3 is ok
+// This function is not in use meanwhile (not verified), as
+// OSD clusters are allowed to access that registry by default
+func configure_pull_secret() error {
+	log.Info("configure_pull_secret()")
+	command := "oc secrets link default redhatio --for=pull"
+	err := run_shell_command(command, "configure_pull_secret()")
+	return err
+}
+
+// Get 3scale admin token (SOP item 10)
+func get_3scale_admin_token() (string, error) {
+	log.Info("get_3scale_admin_token()")
+	command := "oc get secret system-seed -n redhat-rhoam-3scale --template '{{index .data \"ADMIN_ACCESS_TOKEN\"}}' | base64 -d"
```

Comment:
Hi Pavel. thank you very much for your comment. I will try to fix it, and also will look at samples. At the same time, seems to me (maybe I'm wrong : ) testing "scripts" or applications should be simple as possible (regarding each script). But scripts could be connected in suites, with complex logic. So I think that code could be used in places where system commands or REST APIs (curl) are not available. And we have such place. Step 12 is not automated,  so maybe there we could see how to code it (maybe will need to talk with 3scale dev people, to implement it, to avoid using 3scale admin UI (this is our goal - full automation of the procedure).  But I will review carefully all your comments, and will be happy if we could discuss it also. 
Seems to me (as I wrote before, and it could be wrong) - need use system commands as much as possible, but care about comparability for different OS , as you wrote.
## One more thing, wanted to ask you, if possible, -  I found problem in Step 16 (validation), maybe issue with Routes, ...    I unfortunately needed to switch off this task meanwhile. But I will be happy to return as soon as possible.
Big thanks for your Review and your comments. Sincerely.

Reply:
I'd like also suggest following, if possible (regarding using "oc" or other system commands)  - Do it in steps.
1. Need have "script" working reliable, and easy for usage and Understanding, for users, who is working with  H24* procedure, maybe SRE, I don't know. It should be readable, not only for programmers.
2. Use GO libraries meanwhile just in places where we can't do it with system commands. First candidate - step 12. We need avoid using 3scale admin UI.
3. Improvement stage - create GO code project, discuss project structure, split code on small files (not like script) to be easy supportable.
4. Think about Suite of H* tests, about UI implementation (don't know, ... maybe REACT..., although I didn't use it, but worked before with UI in C# and MFC/C++ : )

Reply:
Hi Pavel. Yes, sure, I will check all. Just unfortunately I will be able to work on this only in testing week, but will try find time also before.  Maybe we could also discuss using "oc" and other system commands. Thank you very much for your comment and help

Reply:
Hi Pavel, refactoring done for get_3scale_admin_token () , similar as in your example.
If you will see more candidates for refactoring, it will be appreciated.
Maybe you could help with step 12, that remains manual. Maybe we need consult with 3scale Team ?